### PR TITLE
Token-based authentication without openssl

### PR DESCRIPTION
**Pull request by boxdot**
Add two feature flags: openssl (default) and ring (optional). The
ring feature flags implements token-based authentication without
dependency on openssl. However, it does not support
PKCS#12-certificated authentication.

When both features are enabled: `openssl` is chosen over `ring`. 
At least one of the two features has to be enabled.

MSRV 1.60 (due to the feature section in Cargo.toml using `dep:`)

### CODE REVIEW

File: Cargo.toml
Comment (boxdot) [inline Cargo.toml:19]: maybe it actually makes sence to remove this feature and use the negative feature selector `#[cfg(not(feature = "openssl"))]`. Then dependency on `ring` becomes an impl detail and we don't have to require MSRV 1.60.

**Comment by HarryET** [comment]
👋🏼  Hello, this project now has new maintainers and over come coming week or so I'm going to do a full dive into each open PR/Issue. Is there the use-case of needing token based auth without openssl?

**Comment by HarryET** [comment]
> > Is there the use-case of needing token based auth without openssl?
> 
> openssl is a quite heavy dependency. I understand that not everybody tries to avoid it, but for some project it is too heavy. In my case, it is replaced by rustls (for TLS) and ring (for cryptography primitives).

I see now, yeah could be a good thing to merge in. Any chance you can change this to target the v0.7 branch? and I'll get a review to you asap.

File: Cargo.toml
Comment (boxdot) [reply Cargo.toml:21]: Actually this is not possible. We need a dedicated feature flag to enable crates.

File: Cargo.toml
Comment (HarryET) [inline Cargo.toml:36]: Not needed, please see other comment.

File: src/client.rs
Comment (HarryET) [inline src/client.rs:69]: Why?

File: src/client.rs
Comment (HarryET) [inline src/client.rs:77]: Please change to use `cfg!()` macro

File: src/signer.rs
Comment (HarryET) [inline src/signer.rs:85]: Please change to use `cfg!()` macro

**Review by HarryET — CHANGES_REQUESTED** [review]
Heya, here is a review with some comments that I have from an initial read through. Will try and get another review for you too.

**Comment by HarryET** [comment]
Hey @boxdot, please review the comments left by myself and @ennmichael. Also please fix the Clippy errors.

**Comment by boxdot** [comment]
> Thanks for your PR! Hopefully some of this feedback is constructive.

Thank you. It is very helpful.

File: Cargo.toml
Comment (boxdot) [reply Cargo.toml:36]: https://github.com/rust-lang/cargo/issues/1839

File: Cargo.toml
Comment (boxdot) [reply Cargo.toml:21]: https://github.com/rust-lang/cargo/issues/1839

File: Cargo.toml
Comment (boxdot) [reply Cargo.toml:36]: Removed.

File: src/client.rs
Comment (boxdot) [reply src/client.rs:69]: With removal of cfg_if and by making openssl the default one not needed anymore.

File: src/client.rs
Comment (boxdot) [reply src/client.rs]: Disabled the whole method for `ring`.

**Comment by boxdot** [comment]
Please note, to enable ring, openssl has to be disable now. And at least one of the two features has to be enabled (`--no-default-features` does not compile due to explicit compilation error).

**Comment by HarryET** [comment]
> Please note, to enable ring, openssl has to be disable now. And at least one of the two features has to be enabled (`--no-default-features` does not compile due to explicit compilation error).

Just to clarify if both are enabled it will still work, correct?

File: src/signer.rs
Comment (boxdot) [reply src/signer.rs:217]: `ring` has a trait `SecureRandom`, which is implemented by https://docs.rs/ring/0.16.20/ring/rand/struct.SystemRandom.html. They use a secure system-provided random generator.

However, now reading the docs I see two other problems:

1. I used a deprecated method `sign`,
2. Docs of `ring` suggest not to create `SecureRandom` all the time, but instead pass it as param.

**Comment by boxdot** [comment]
> 



> Just to clarify if both are enabled it will still work, correct?

Yes, if both are enabled, `openssl` takes precedence and is used.

File: src/signer.rs
Comment (boxdot) [reply src/signer.rs]: Good idea. The random generator is now stored in the `Secret`. I asked about the second point on `ring`'s issue tracker. It might take a while until they answer. I don't see how not to use this method atm.

File: README.md
Comment (HarryET) [inline README.md:11]: What is the reason for the version bump in this case?

**Comment by HarryET** [comment]
Just going to wait for a final review from @ennmichael and maybe another person from WalletConnect before merge

File: README.md
Comment (boxdot) [reply README.md:11]: The "dep:" syntax in features is only supported starting with 1.60 (https://doc.rust-lang.org/cargo/reference/features.html)

**Comment by HarryET** [comment]
Just to keep you updated, I'm away from work for a week or so but will try to get a review from someone else for you

File: README.md
Comment (HarryET) [reply README.md:11]: 👍🏼 cool no worries then

**Review by arein — APPROVED** [review]
LGTM - only small request - can you update the `Features` section in the `README` to reflect this before pushing pls?

**Comment by boxdot** [comment]
> LGTM - only small request - can you update the `Features` section in the `README` to reflect this before pushing pls?

Updated

**Comment by HarryET** [comment]
Will review this on Monday and merge if all looks good, thanks again for the PR and sorry for the delay.

**Comment by pimeys** [comment]
Maybe one thing that could be added here is to get GitHub actions running without OpenSSL and using ring, at least clippy/format/compilation to see nothing breaks in the future.

**Comment by boxdot** [comment]
> Maybe one thing that could be added here is to get GitHub actions running without OpenSSL and using ring, at least clippy/format/compilation to see nothing breaks in the future.

I added a jobs to run clippy and tests with the ring feature enabled only

**Review by HarryET — APPROVED** [review]
Thanks again for the PR! Sorry for the delay, going to merge it in to v0.7 and should be released soon.

**Comment by boxdot** [comment]
Thank you for reviewing.

JFYI: I don't have permissions to merge it btw.

**Comment by HarryET** [comment]
> Thank you for reviewing.
> 
> JFYI: I don't have permissions to merge it btw.

Yeah, just a bit of a delay in merging from my end. Going to do it now; thanks again for the contributions.

**Pull request merged**
